File: backend-on-python/app/modules/auth/auth_service.py
from datetime import datetime, timedelta
from typing import Optional

from fastapi import HTTPException, status, Response
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config.config import get_settings
from app.core.models.users_model import User, UserSession, SubscriptionType
from app.core.security.password import hash_password, verify_password
from app.core.utils.auth_utils import create_session_token, set_auth_cookies, clear_auth_cookies
from app.modules.auth.auth_repository import AuthRepository
from app.modules.auth.auth_schemas import (
    UserSignupRequest,
    UserLoginRequest,
    UserResponse,
    TokenResponse,
    CurrentUserResponse,
    UserOnboardingRequest,
)


class AuthService:
    def __init__(self, session: AsyncSession):
        self.repository = AuthRepository(session)
        self.settings = get_settings()

    async def signup_user(self, response: Response, user_data: UserSignupRequest) -> TokenResponse:
        existing_user = await self.repository.get_user_by_email_or_username(user_data.email)
        if existing_user:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="User with this email or username already exists",
            )

        hashed_password = hash_password(user_data.password)
        user = await self.repository.create_user(user_data, hashed_password)
        session_token = create_session_token(user.id, user.role, user.subscription)
        expires_at = datetime.utcnow() + timedelta(
            seconds=self.settings.security.jwt_session_token_expire_secs
        )
        await self.repository.create_or_update_user_session(user.id, session_token, expires_at)
        set_auth_cookies(response, session_token, expires_at)
        return TokenResponse(
            sessionToken=session_token,
            expiresAt=expires_at,
        )

    async def login_user(self, response: Response, user_data: UserLoginRequest) -> TokenResponse:
        user = await self.repository.get_user_by_email_or_username(user_data.emailOrUsername)
        if not user or not verify_password(user_data.password, user.password):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid credentials",
            )

        session_token = create_session_token(user.id, user.role, user.subscription)
        expires_at = datetime.utcnow() + timedelta(
            seconds=self.settings.security.jwt_session_token_expire_secs
        )

        await self.repository.create_or_update_user_session(user.id, session_token, expires_at)
        set_auth_cookies(response, session_token, expires_at)

        return TokenResponse(
            sessionToken=session_token,
            expiresAt=expires_at,
        )

    async def logout_user(self, response: Response, token: str):
        user_session = await self.repository.get_user_session_by_token(token)
        if user_session:
            await self.repository.delete_user_session(user_session)
        clear_auth_cookies(response)

    async def get_current_user(self, user_id: int) -> CurrentUserResponse:
        user = await self.repository.get_user_by_id(user_id)
        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found",
            )
        return CurrentUserResponse(
            id=user.id,
            firstName=user.first_name,
            lastName=user.last_name,
            email=user.email,
            username=user.username,
            roles=[user.role], # Assuming single role for now
            verified=user.verified,
            onboarded=user.onboarded,
            avatarUrl=None, # Placeholder, implement avatar logic later
        )

    async def onboard_user(self, user_id: int, onboarding_data: UserOnboardingRequest) -> UserResponse:
        user = await self.repository.get_user_by_id(user_id)
        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found",
            )
        updated_user = await self.repository.update_user_onboarding(
            user,
            onboarding_data,
        )
        return UserResponse.model_validate(updated_user)

File: src/modules/auth/auth.service.ts
import {
	Injectable,
	UnauthorizedException,
	ConflictException,
	NotFoundException,
} from '@nestjs/common'
import { PrismaService } from '../../prisma/prisma.service'
import {
	UserSignupRequest,
	UserLoginRequest,
	CurrentUserResponse,
	UserOnboardingRequest,
} from './dtos/auth.dtos'
import { CookieService } from '../../common/services/cookie.service'
import { hash, compare } from 'bcryptjs'
import { Response } from 'express'
import { UserRole, SubscriptionType } from '@prisma/client'
import { UserSessionService } from './services/session.service'
import { S3Service } from '../../common/s3/s3.service'
import { MediaUrlService } from '../../common/services/media-url.service'
import { Express } from 'express' // Import Express for Multer types

const SESSION_TOKEN_EXPIRE_SECS = 60 * 60 * 24 * 30 // Example: 30 days

@Injectable()
export class AuthService {
	constructor(
		private readonly prisma: PrismaService,
		private readonly cookieService: CookieService,
		private readonly userSessionService: UserSessionService,
		private readonly s3Service: S3Service,
		private readonly mediaUrlService: MediaUrlService, // Inject MediaUrlService
	) {}

	async signupUser(response: Response, userData: UserSignupRequest): Promise<void> {
		const existingUser = await this.prisma.user.findFirst({
			where: {
				OR: [{ email: userData.email }, { username: userData.username }],
			},
		})
		if (existingUser) {
			throw new ConflictException('User with this email or username already exists')
		}

		const hashedPassword = await hash(userData.password, 10)
		const user = await this.prisma.user.create({
			data: {
				firstName: userData.firstName,
				lastName: userData.lastName,
				email: userData.email,
				username: userData.username,
				password: hashedPassword,
				subscription: SubscriptionType.FREE,
				role: UserRole.USER,
				verified: false,
				onboarded: false,
			},
		})

		const token = await this.userSessionService.upsertSession(user.id)
		this.cookieService.setAuthCookie(response, token, SESSION_TOKEN_EXPIRE_SECS * 1000)

		return
	}

	async loginUser(response: Response, userData: UserLoginRequest): Promise<void> {
		const user = await this.prisma.user.findFirst({
			where: {
				OR: [{ email: userData.emailOrUsername }, { username: userData.emailOrUsername }],
			},
		})
		if (!user || !(await compare(userData.password, user.password))) {
			throw new UnauthorizedException('Invalid credentials')
		}

		const token = await this.userSessionService.upsertSession(user.id)
		this.cookieService.setAuthCookie(response, token, SESSION_TOKEN_EXPIRE_SECS * 1000)

		return
	}

	async logoutUser(response: Response, sessionToken: string): Promise<void> {
		await this.userSessionService.revokeUserSession(sessionToken)
		this.cookieService.clearAuthCookie(response)
	}

	async uploadAvatar(userId: number, file: Express.Multer.File): Promise<CurrentUserResponse> {
		const user = await this.prisma.user.findUnique({ where: { id: userId } })
		if (!user) {
			throw new NotFoundException('User not found')
		}

		const fileExtension = file.originalname.split('.').pop()
		const avatarKey = `avatars/${userId}/avatar.${fileExtension}`

		await this.s3Service.uploadObject(avatarKey, file.buffer, file.mimetype)

		const updatedUser = await this.prisma.user.update({
			where: { id: userId },
			data: { avatarKey },
		})

		return {
			id: updatedUser.id,
			firstName: updatedUser.firstName,
			lastName: updatedUser.lastName,
			email: updatedUser.email,
			username: updatedUser.username,
			role: updatedUser.role,
			verified: updatedUser.verified,
			onboarded: updatedUser.onboarded,
			avatarUrl: updatedUser.avatarKey
				? MediaUrlService.buildMediaUrl(updatedUser.avatarKey)
				: null,
		}
	}

	async uploadBanner(userId: number, file: Express.Multer.File): Promise<CurrentUserResponse> {
		const user = await this.prisma.user.findUnique({ where: { id: userId } })
		if (!user) {
			throw new NotFoundException('User not found')
		}

		const fileExtension = file.originalname.split('.').pop()
		const bannerKey = `banners/${userId}/banner.${fileExtension}`

		await this.s3Service.uploadObject(bannerKey, file.buffer, file.mimetype)

		const updatedUser = await this.prisma.user.update({
			where: { id: userId },
			data: { bannerKey },
		})

		return {
			id: updatedUser.id,
			firstName: updatedUser.firstName,
			lastName: updatedUser.lastName,
			email: updatedUser.email,
			username: updatedUser.username,
			role: updatedUser.role,
			verified: updatedUser.verified,
			onboarded: updatedUser.onboarded,
			avatarUrl: updatedUser.avatarKey
				? MediaUrlService.buildMediaUrl(updatedUser.avatarKey)
				: null,
			bannerUrl: updatedUser.bannerKey
				? MediaUrlService.buildMediaUrl(updatedUser.bannerKey)
				: null,
		}
	}

	async getCurrentUser(userId: number): Promise<CurrentUserResponse> {
		const user = await this.prisma.user.findUnique({ where: { id: userId } })
		if (!user) {
			throw new NotFoundException('User not found')
		}
		return {
			id: user.id,
			firstName: user.firstName,
			lastName: user.lastName,
			email: user.email,
			username: user.username,
			role: user.role,
			verified: user.verified,
			onboarded: user.onboarded,
			avatarUrl: user.avatarKey ? MediaUrlService.buildMediaUrl(user.avatarKey) : null,
			bannerUrl: user.bannerKey ? MediaUrlService.buildMediaUrl(user.bannerKey) : null,
		}
	}

	async onboardUser(
		userId: number,
		onboardingData: UserOnboardingRequest,
	): Promise<CurrentUserResponse> {
		const user = await this.prisma.user.findUnique({ where: { id: userId } })
		if (!user) {
			throw new NotFoundException('User not found')
		}

		// Update user onboarding fields
		const updatedUser = await this.prisma.user.update({
			where: { id: userId },
			data: {
				cityId: onboardingData.cityId,
				dateOfBirth: onboardingData.dateOfBirth,
				interests: onboardingData.interests,
				organizationName: onboardingData.organizationName,
				bioAbout: onboardingData.about,
				onboarded: true,
				academicInfo: onboardingData.academic
					? {
							upsert: {
								update: { ...onboardingData.academic },
								create: { ...onboardingData.academic },
							},
						}
					: undefined,
				languageProficiencies: {
					deleteMany: {}, // Remove old
					create: onboardingData.languages.map(lang => ({
						language: lang.language,
						level: lang.level,
					})),
				},
			},
		})

		return {
			id: updatedUser.id,
			firstName: updatedUser.firstName,
			lastName: updatedUser.lastName,
			email: updatedUser.email,
			username: updatedUser.username,
			role: updatedUser.role,
			verified: updatedUser.verified,
			onboarded: updatedUser.onboarded,
			avatarUrl: updatedUser.avatarKey
				? MediaUrlService.buildMediaUrl(updatedUser.avatarKey)
				: null,
			bannerUrl: updatedUser.bannerKey
				? MediaUrlService.buildMediaUrl(updatedUser.bannerKey)
				: null,
		}
	}
}
